### ISSUE TITLE
[Bug]: query planner not using index with subquery & multiple chunks

### ISSUE THREAD

**Issue by phemmer** [timescale/timescaledb#7068]
### What type of bug is this?

Performance issue

### What subsystems and features are affected?

Query planner

### What happened?

When I execute a query that uses a subquery filter & multiple chunks, the wrong (or no) index is used, causing a large performance degradation. If I don't use a subquery, or if I only query a single chunk at a time, it works fine.

Here's an example showing the issue:
https://explain.dalibo.com/plan/a5bh7372bgcg0ee8#raw 
[2_chunks_subquery.txt](https://github.com/user-attachments/files/15976155/2_chunks_subquery.txt)

We can see that on `_timescaledb_internal._hyper_2427_264588_chunk`, it's doing a `seq scan` without using an index, takes 10 seconds, and returns 27,604,988 rows, causing a ton of work for the higher operations.
I have an index, which is on both the `tag_id` and `time` columns, which would result in a much faster query. This is why I'm using a materialized CTE here, as I was trying to strongly encourage postgres to use the index containing the `tag_id` column. No matter if I use a normal subquery, a join, etc, none result in using the correct index.

If I manually take that subquery (the CTE), evaluate it, and copy/paste the results into the `where` clause, it goes much faster:
https://explain.dalibo.com/plan/54h8h7b5ee5b8gd6#raw 
[2_chunks_copypaste.txt](https://github.com/user-attachments/files/15976157/2_chunks_copypaste.txt)

We can see now that the correct index was used (`_hyper_2427_264588_chunk_haproxy_server_tag_id_time_idx`), which returned only 2,400 rows, and completed in 3.2ms.

Both the above queries spanned 2 chunks. If I reduce to just the second (chronologically) of the two chunks (the one that resulted in the performance difference in the above 2 queries), though still using the subquery, the plan again uses the correct index:
https://explain.dalibo.com/plan/556dh3acg5f3173g#raw 
[1_chunk_subquery.txt](https://github.com/user-attachments/files/15976159/1_chunk_subquery.txt)

And just for comparison, when using copy/paste instead of subquery, it has similar plan & performance:
https://explain.dalibo.com/plan/825h52f73d389f9h#raw 
[1_chunk_copypaste.txt](https://github.com/user-attachments/files/15976160/1_chunk_copypaste.txt)

So basically:
* 2 chunks (one compressed, one uncompressed) with subquery: bad
* 2 chunks (one compressed, one uncompressed) with copy/paste: good
* 1 chunk (uncompressed) with subquery: good
* 1 chunk (uncompressed) with copy/paste: good

### TimescaleDB version affected

2.14.2

### PostgreSQL version used

16.2

### What operating system did you use?

Debian 16

### What installation method did you use?

Deb/Apt

### What platform did you run on?

On prem/Self-hosted

### How can we reproduce the bug?

```bash
I can try to create a reproduction if necessary, but would be a fair amount of work, and I wanted to at least dump the information acquired so far.
```

**Comment by nikkhils** [comment]
@phemmer the main issue is that it's not able to push the materialized subquery down to the filter level for the scans on the chunks to be effective. This might be a core PostgreSQL planner limitation.

**Comment by phemmer** [comment]
Ok, I've managed to create a reproducer. It has to do with compression. If the first of the 2 chunks is compressed, then a seq scan is used on the second chunk. But if the first chunk is decompressed, an index scan is used on both.

**Comment by nikkhils** [comment]
@phemmer care to share the reproducer script here?

**Comment by phemmer** [comment]
I already did. it's in the description

**Comment by akuzm** [comment]
Yeah, it seems to be unable to generate a parameterized path on compressed chunks, this is the closest I can get, with some modifications to nudge the hypertable to the inner side of the nested loop:
```
set enable_hashjoin to off; set enable_mergejoin to off;

explain (analyze, costs)
with tag_ids as materialized (select distinct generate_series(0,29) as tag_id)
select tag_id, a from tag_ids left join test using (tag_id)
where time between '2024-06-25T00:00:00Z' and '2024-06-25T07:59:59Z'
;
                                                                                   QUERY PLAN                                                                                    
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Nested Loop  (cost=0.84..2554411.22 rows=720000 width=12) (actual time=0.831..13072.552 rows=14400 loops=1)
   Join Filter: (tag_ids.tag_id = _hyper_25_18349_chunk.tag_id)
   Rows Removed by Join Filter: 143985600
   CTE tag_ids
     ->  HashAggregate  (cost=0.24..0.62 rows=30 width=4) (actual time=0.007..0.065 rows=30 loops=1)
           Group Key: generate_series(0, 29)
           Batches: 1  Memory Usage: 24kB
           ->  ProjectSet  (cost=0.00..0.17 rows=30 width=4) (actual time=0.001..0.003 rows=30 loops=1)
                 ->  Result  (cost=0.00..0.01 rows=1 width=0) (actual time=0.000..0.001 rows=1 loops=1)
   ->  CTE Scan on tag_ids  (cost=0.00..0.60 rows=30 width=4) (actual time=0.008..0.121 rows=30 loops=1)
   ->  Append  (cost=0.23..25087.00 rows=4800000 width=12) (actual time=0.007..311.364 rows=4800000 loops=30)
         ->  Custom Scan (DecompressChunk) on _hyper_25_18349_chunk  (cost=0.23..545.00 rows=2400000 width=12) (actual time=0.007..75.672 rows=2400000 loops=30)
               Vectorized Filter: (("time" >= '2024-06-25 02:00:00+02'::timestamp with time zone) AND ("time" <= '2024-06-25 09:59:59+02'::timestamp with time zone))
               ->  Seq Scan on compress_hyper_26_18353_chunk  (cost=0.00..545.00 rows=2400 width=116) (actual time=0.002..0.841 rows=2400 loops=30)
                     Filter: ((_ts_meta_max_1 >= '2024-06-25 02:00:00+02'::timestamp with time zone) AND (_ts_meta_min_1 <= '2024-06-25 09:59:59+02'::timestamp with time zone))
         ->  Custom Scan (DecompressChunk) on _hyper_25_18350_chunk  (cost=0.23..542.00 rows=2400000 width=12) (actual time=0.007..72.120 rows=2400000 loops=30)
               Vectorized Filter: (("time" >= '2024-06-25 02:00:00+02'::timestamp with time zone) AND ("time" <= '2024-06-25 09:59:59+02'::timestamp with time zone))
               ->  Seq Scan on compress_hyper_26_18354_chunk  (cost=0.00..542.00 rows=2400 width=116) (actual time=0.002..0.726 rows=2400 loops=30)
                     Filter: ((_ts_meta_max_1 >= '2024-06-25 02:00:00+02'::timestamp with time zone) AND (_ts_meta_min_1 <= '2024-06-25 09:59:59+02'::timestamp with time zone))
```
